src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -686,6 +688,37 @@ private async ValueTask SendRstStreamAsync(int streamId, Http2ProtocolErrorCode
             }
         }
 
+        private async ValueTask SendGoAwayAsync(int lastStreamId, Http2ProtocolErrorCode errorCode, ReadOnlyMemory<byte> additionalDebugData)
+        {
+            await AcquireWriteLockAsync().ConfigureAwait(false);
+            try
+            {
+                int length = FrameHeader.Size + FrameHeader.GoAwayMinLength + additionalDebugData.Length;
+                _outgoingBuffer.EnsureAvailableSpace(length);
+                WriteFrameHeader(new FrameHeader(length, FrameType.GoAway, FrameFlags.None, 0));
+
+                _outgoingBuffer.AvailableSpan[0] = (byte)(((int)lastStreamId & 0xFF000000) >> 24);
+                _outgoingBuffer.AvailableSpan[1] = (byte)(((int)lastStreamId & 0x00FF0000) >> 16);
+                _outgoingBuffer.AvailableSpan[2] = (byte)(((int)lastStreamId & 0x0000FF00) >> 8);
+                _outgoingBuffer.AvailableSpan[3] = (byte)((int)lastStreamId & 0x000000FF);
+
+                _outgoingBuffer.AvailableSpan[4] = (byte)(((int)errorCode & 0xFF000000) >> 24);
+                _outgoingBuffer.AvailableSpan[5] = (byte)(((int)errorCode & 0x00FF0000) >> 16);
+                _outgoingBuffer.AvailableSpan[6] = (byte)(((int)errorCode & 0x0000FF00) >> 8);
+                _outgoingBuffer.AvailableSpan[7] = (byte)((int)errorCode & 0x000000FF);
```

Comment:
Rather than doing all of this manually, can we instead use the appropriate methods from BinaryPrimitives, e.g. BinaryPrimitives.WriteInt32BigEndian?

Reply:
That cleaned things up significantly. Thanks.